### PR TITLE
Add SDFormat model plugins

### LINKED ISSUE #610: RobotImporter incorrectly adding Robot Control component to non-robot imports
**Describe the bug**
`RobotControl` component is added to the first link of the first model while importing data from URDF/SDFormat. This is an incorrect behavior for non-robot imports. In the previous release, _RobotImporter_ was used to import URDF files with a single robot only. This is not the case anymore, as _RobotImporter_ can parse SDFormat data with multiple models (robots) or even worlds. E.g. `RobotControl` component is added to _air conditioning_ link when importing [aws-bookstore-world](https://github.com/aws-robotics/aws-robomaker-bookstore-world).

It could be also considered to rename _RobotImporter_ into something more generic.

**Expected behavior**
Do not add `RobotControl` component when not needed.

**Actual behavior**
`RobotControl` component is added to the first link of the first model.

### PR DESCRIPTION
**Pull request by jhanca-robotecai**
- Closes #610 
- Add hook template for SDFormat model plugins
- Add pipeline for processing hooks
- Add a sample hook for `libgazebo_ros_skid_steer_drive.so` and `libgazebo_ros_diff_drive`

The code was tested using modified _ROSBot_ urdf files. The only modification was adding the plugin itself based on [libgazebo_ros_skid_steer_drive docs](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-Skid-Steer-drive) and [libgazebo_ros_diff_drive docs](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-Migration:-Diff-drive). Please find the robots [attached](https://github.com/o3de/o3de-extras/files/13447959/test_models.zip).

~~The attached robots [test_models.zip](https://github.com/o3de/o3de-extras/files/13447959/test_models.zip) require one change after import: setting "Use motor" toggle to true in all used wheels. This part will be fixed in the next PR.~~

Please note, that ROS topic names are not parsed at this point, therefore the robots can be controlled via `/base_link/default_topic` topic. This will be fixed in another PR (topic names are not parsed in Sensor importer either).

### CODE REVIEW

**Comment by jhanca-robotecai** [comment]
The current version requires changes in o3de, please see https://github.com/o3de/o3de/pull/17124

It works with both articulations and joints. Note, that the torque in case of articulations is too high and the imported samples fly away. This can be changed e.g. by modifying the weight of a _base_link_.

File: Gems/ROS2/Code/Include/ROS2/RobotImporter/SDFormatModelPluginImporterHook.h
Comment (adamdbrw) [reply Gems/ROS2/Code/Include/ROS2/RobotImporter/SDFormatModelPluginImporterHook.h:34]: I know this is not uniform in our codebase yet, but when writing about the Gem and components in a descriptive manner, use space in "ROS 2"

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:37]: Is it possible for the entityResult to not be success here?

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:42]: Nit: return early instead.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:49]: Switching to early return here will improve readability.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:257]: Note: good to think what to abstract here as ackermann drive support will be added to avoid any copying.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2SDFormatHooksUtils.h
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2SDFormatHooksUtils.h:24]: "ROS 2" here and overall

**Review by adamdbrw — COMMENTED** [review]
Good work! I would like this feature to be merged once documentation PR (at least a draft) is submitted about hooks and their registration mechanism. This would help to review the PR. 

The documentation can be based on RFC text.

File: Gems/ROS2/Code/Include/ROS2/RobotImporter/SDFormatModelPluginImporterHook.h
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Include/ROS2/RobotImporter/SDFormatModelPluginImporterHook.h:34]: Thanks for the comment, I will modify my current uses and change all incorrect usages whenever I touch the code around one.

File: Gems/ROS2/Code/Source/RobotImporter/ROS2RobotImporterEditorSystemComponent.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/ROS2RobotImporterEditorSystemComponent.cpp:47]: I added the description for SDFormat sensor importer with explanations about this part. I will duplicate it in the model plugin importer description when it passes the review: https://github.com/o3de/o3de.org/pull/2515

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:28]: I realized the same when implementing _Ackermann_ control importer. Moved to _Utils_.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:33]: No, but this line gets the name of the link:
parent link -> joint -> child link

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:37]: It should not happen, but we had segmentation faults in similar cases due to name collisions lately (name collisions leading to not creating all entities). It is always safer to add a check.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:49]: My idea was to keep _hooks_ independent from the rest of the code. I do not want to pass too many parameters to the creation callback either to keep the code simple and easy to implement in external Gems. This means there is no information in the _hook_  if the importer uses the articulations or not.

The current flow is following:
1) assume _no articulation_ and set _enable motor_ in joint; terminate if succeed
2) assume _articulation_ and add _WheelComponent_ required in _use articulation_ configuration; terminate if succeed
3) print warning if both failed

I can move the warning and early return before, but it would still require an early return after 1)

File: Gems/ROS2/Code/Source/RobotImporter/URDF/ArticulationsMaker.cpp
Comment (michalpelka) [inline Gems/ROS2/Code/Source/RobotImporter/URDF/ArticulationsMaker.cpp:58]: Consider : AZ::Vector3::CreateAxisX() and removing `o3deJointDir`

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2ModelPluginHooks.h
Comment (michalpelka) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2ModelPluginHooks.h:13]: Consider : ROS2::SDFormat::ROS2ModelPluginHooks

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2SDFormatHooksUtils.cpp
Comment (michalpelka) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2SDFormatHooksUtils.cpp:53]: I'm not too fond of activation to change something in the component state.
PhysX::EditorHingeJointComponent is dependent on service : `PhysicsDynamicRigidBodyService`
Please consider :
- Checking after activation if the component is **really** active and performing early return and error message.
- Call the Handler interface directly without activation, looking at the ` EditorHingeJointComponent::SetBoolValue` implementation, it is nothing more than a setter.
Btw is component EditorHingeJointComponent active at this state ?

File: Gems/ROS2/Code/Source/RobotImporter/Utils/RobotImporterUtils.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/Utils/RobotImporterUtils.cpp:1062]: I tend to apply _clang-format_ to every file I edit, which often modifies parts of the file I did not change myself. On one hand, it increases the diffs in PRs, but on the other hand, it allows for cleaning up the incorrectly formatted code step by step.

I understand this PR introduces too many changes, so I reverted most of the changes caused by _clang-format_ from this file.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2ModelPluginHooks.h
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/ROS2ModelPluginHooks.h]: Fixed.

File: Gems/ROS2/Code/Source/RobotImporter/URDF/ArticulationsMaker.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/URDF/ArticulationsMaker.cpp]: Fixed.

File: Gems/ROS2/Code/Source/RobotImporter/ROS2RobotImporterEditorSystemComponent.h
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/ROS2RobotImporterEditorSystemComponent.h:52]: Could you document the return value of this method.
It return true if the attribute is not found.
However if the attribute was found, but not successfully read it return `false`

That seems strange to me.
It seems like it should return true if the attribute was found and was successfully read.

File: Gems/ROS2/Code/Source/RobotImporter/URDF/URDFPrefabMaker.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/URDF/URDFPrefabMaker.cpp:473]: This comment is now out of date.

All of the links which aren't a child of another link would have a control plugin added to it.

Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/URDF/URDFPrefabMaker.cpp]: Fixed.

File: Gems/ROS2/Code/Source/RobotImporter/ROS2RobotImporterEditorSystemComponent.h
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/ROS2RobotImporterEditorSystemComponent.h]: This method is used as a callback in `EnumerateAll`. In this case `EnumerateAll` does not care if the attribute was found or not, but it cares if the search should _continue_ or not. Hence the inverted condition: the search should _not_ continue if the attribute was found.

But you are right I made a mistake here. This callback should always return `true` to ensure the search always continues, as the same attribute can be used in multiple classes declaring multiple hooks and, therefore, `EnumerateAll` should not stop after finding the first instance.

I added comments to explain it in detail.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:26]: Nit: looking at what it does, it should probably be named CreateVehicleConfiguration

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:50]:
```suggestion
                        "CreateVehicleConfiguration",
```

Comment (adamdbrw) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:50]: If the name is updated. In any case, this is a util so the source should be changed.

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:116]: Change the source string accordingly, see comment above. Applies to other places.

File: Gems/ROS2/Code/Source/VehicleDynamics/ModelLimits/SkidSteeringModelLimits.cpp
Comment (adamdbrw) [inline Gems/ROS2/Code/Source/VehicleDynamics/ModelLimits/SkidSteeringModelLimits.cpp:70]: This seems to do the same as SetAngularAccelerationLimit member function. Call it here, or remove the function if it is not meant to be used in isolation (limits are set together each time).

**Review by adamdbrw — COMMENTED** [review]
Valuable contribution, some comments (rather minor).

File: Gems/ROS2/Code/Source/VehicleDynamics/ModelLimits/SkidSteeringModelLimits.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/VehicleDynamics/ModelLimits/SkidSteeringModelLimits.cpp]: I removed the constructor as AngularAccelerationLimit is the only limit I set.

**Review by michalpelka — APPROVED** [review]

**Review by michalpelka — APPROVED** [review]

**Comment by michalpelka** [comment]
Tested against development.

**Review by adamdbrw — APPROVED** [review]

**Comment by jhanca-robotecai** [comment]
Rebased and force-pushed on top of `development` to rerun and fix Jenkins output.

**Comment by jhanca-robotecai** [comment]
Rebased and force-pushed again after #641 was merged

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (spham-amzn) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:54]: More detailed warnings is always helpful.

```suggestion
                    AZ_Warning(
                        "CreateVehicleConfiguration",
                        entityIdLeft.IsValid(),
                        "Cannot find entity ID for one of the joint %s",
                        jointNameLeft.c_str());
                    AZ_Warning(
                        "CreateVehicleConfiguration",
                        entityIdRight.IsValid(),
                        "Cannot find entity ID for one of the joints %s",
                        jointNameRight.c_str());
```

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (spham-amzn) [inline Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:117]: FYI: There is an AZCore math function that does this as well (you'll need to include <AzCore/Math/MathUtils.h>

```suggestion
                            AZ::IsClose(configuration.m_wheelbase, wheelSeparation->Get<float>(), epsilon),
```

File: Gems/ROS2/Code/Source/VehicleDynamics/ModelComponents/SkidSteeringModelComponent.cpp
Comment (spham-amzn) [inline Gems/ROS2/Code/Source/VehicleDynamics/ModelComponents/SkidSteeringModelComponent.cpp:46]: why not move `m_vehicleConfiguration` into the initialization list above?

**Review by spham-amzn — APPROVED** [review]

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:54]: I will modify the code accordingly in the following PR (Ackermann control), as the automated testing already succeeded in this PR and this PR was on hold for a long time.

Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:54]: Pushed in https://github.com/o3de/o3de-extras/pull/642

File: Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp
Comment (jhanca-robotecai) [reply Gems/ROS2/Code/Source/RobotImporter/SDFormat/Hooks/ROS2SkidSteeringModelHook.cpp:117]: Pushed in https://github.com/o3de/o3de-extras/pull/642

File: Gems/ROS2/Code/Source/VehicleDynamics/ModelComponents/SkidSteeringModelComponent.cpp
Comment (spham-amzn) [reply Gems/ROS2/Code/Source/VehicleDynamics/ModelComponents/SkidSteeringModelComponent.cpp:46]: Too bad there isn't a constructor in the base class to pass this in.

**Pull request merged**
